src/core/server/saved_objects/export/saved_objects_exporter.ts

```diff
@@ -0,0 +1,154 @@
+  SavedObjectsExportByTypeOptions,
+} from './types';
+
+/**
+ * @public
+ */
+export type ISavedObjectsExporter = PublicMethodsOf<SavedObjectsExporter>;
+
+/**
+ * @public
+ */
+export class SavedObjectsExporter {
+  private savedObjectsClient: SavedObjectsClientContract;
+  private exportSizeLimit: number;
+
+  constructor({
+    savedObjectsClient,
+    exportSizeLimit,
+  }: {
+    savedObjectsClient: SavedObjectsClientContract;
+    exportSizeLimit: number;
+  }) {
+    this.savedObjectsClient = savedObjectsClient;
+    this.exportSizeLimit = exportSizeLimit;
+  }
+
+  /**
+   * Generates an export stream for given types.
+   *
+   * See the {@link SavedObjectsExportByTypeOptions | options} for more detailed information.
+   */
+  public async exportByTypes(options: SavedObjectsExportByTypeOptions) {
+    const objects = await this.fetchByTypes(options);
+    return this.processObjects(objects, options);
+  }
+
+  /**
+   * Generates an export stream for given object references.
+   *
+   * See the {@link SavedObjectsExportByObjectOptions | options} for more detailed information.
+   */
+  public async exportByObjects(options: SavedObjectsExportByObjectOptions) {
+    if (options.objects.length > this.exportSizeLimit) {
+      throw Boom.badRequest(`Can't export more than ${this.exportSizeLimit} objects`);
+    }
+    const objects = await this.fetchByObjects(options);
+    return this.processObjects(objects, options);
+  }
+
+  private async processObjects(
+    rootObjects: SavedObject[],
+    {
+      excludeExportDetails = false,
+      includeReferencesDeep = false,
+      namespace,
+    }: SavedObjectExportBaseOptions
+  ) {
+    let exportedObjects: Array<SavedObject<unknown>> = [];
+    let missingReferences: SavedObjectsExportResultDetails['missingReferences'] = [];
+
+    if (includeReferencesDeep) {
+      const fetchResult = await fetchNestedDependencies(
+        rootObjects,
+        this.savedObjectsClient,
+        namespace
+      );
+      exportedObjects = sortObjects(fetchResult.objects);
+      missingReferences = fetchResult.missingRefs;
+    } else {
+      exportedObjects = sortObjects(rootObjects);
+    }
+
+    // redact attributes that should not be exported
+    const redactedObjects = exportedObjects.map<SavedObject<unknown>>(
+      ({ namespaces, ...object }) => object
+    );
+
+    const exportDetails: SavedObjectsExportResultDetails = {
+      exportedCount: exportedObjects.length,
+      missingRefCount: missingReferences.length,
+      missingReferences,
+    };
+    return createListStream([...redactedObjects, ...(excludeExportDetails ? [] : [exportDetails])]);
+  }
+
+  private async fetchByObjects({ objects, namespace }: SavedObjectsExportByObjectOptions) {
+    const bulkGetResult = await this.savedObjectsClient.bulkGet(objects, { namespace });
+    const erroredObjects = bulkGetResult.saved_objects.filter((obj) => !!obj.error);
+    if (erroredObjects.length) {
+      const err = Boom.badRequest();
```

Comment:
question: the code assumes runtime context being Hapi request handler. Can we throw a custom error instead of `Boom` and adopt the caller's code?

Reply:
Yea, I wanted to get rid of those, but I was afraid of introducing some behavioral changes in the `spaces` plugin routes using the import/exporter. I guess now is still the correct time to do that.

Reply:
Attempted that in https://github.com/elastic/kibana/pull/86264/commits/cffcd0b90dcadf75c3ef98fbab85af332fa60d35. We'll see what the CI thinks about it